st2exporter/st2exporter/worker.py

```diff
@@ -13,118 +13,65 @@
+
         if wait:
             self.wait()
 
-    def wait(self):
-        self._consumer_thread.wait()
-        self._dumper.wait()
-
     def shutdown(self):
-        self._dumper.stop()
-        super(ExecutionsExporter, self).shutdown()
-
-    def process(self, execution):
-        LOG.debug('Got execution from queue: %s', execution)
-        if execution.status not in COMPLETION_STATUSES:
-            return
-        execution_api = ActionExecutionAPI.from_model(execution, mask_secrets=True)
-        self.pending_executions.put_nowait(execution_api)
-        LOG.debug("Added execution to queue.")
-
-    def _bootstrap(self):
-        marker = self._get_export_marker_from_db()
-        LOG.info('Using marker %s...' % marker)
-        missed_executions = self._get_missed_executions_from_db(export_marker=marker)
-        LOG.info('Found %d executions not exported yet...', len(missed_executions))
-
-        for missed_execution in missed_executions:
-            if missed_execution.status not in COMPLETION_STATUSES:
-                continue
-            execution_api = ActionExecutionAPI.from_model(missed_execution, mask_secrets=True)
+        for model_exporter in self._model_exporters:
+            while not self._model_exporters_pool.free():
+                eventlet.sleep(0.1)
             try:
-                LOG.debug('Missed execution %s', execution_api)
-                self.pending_executions.put_nowait(execution_api)
+                self._model_exporters_pool.spawn(model_exporter.shutdown)
             except:
-                LOG.exception('Failed adding execution to in-memory queue.')
-                continue
-        LOG.info('Bootstrapped executions...')
+                LOG.exception('Failed shutting down model exporter %s.', model_exporter)
+        self._stopped = True
 
-    def _get_export_marker_from_db(self):
-        try:
-            markers = DumperMarker.get_all()
-        except:
-            return None
-        else:
-            if len(markers) >= 1:
-                marker = markers[0]
-                return isotime.parse(marker.marker)
-            else:
-                return None
-
-    def _get_missed_executions_from_db(self, export_marker=None):
-        if not export_marker:
-            return self._get_all_executions_from_db()
-
-        # XXX: Should adapt this query to get only executions with status
-        # in COMPLETION_STATUSES.
-        filters = {'end_timestamp__gt': export_marker}
-        LOG.info('Querying for executions with filters: %s', filters)
-        return ActionExecution.query(**filters)
-
-    def _get_all_executions_from_db(self):
-        return ActionExecution.get_all()  # XXX: Paginated call.
+    def wait(self):
+        for model_exporter in self._model_exporters:
+            model_exporter.wait()
+            LOG.info('Model exporter %s quit.', model_exporter)
+            # XXX: We should probably raising an exception at this point.
```

Comment:
> I have one question on that PR.
I guess if you do not expect an exporter to every quit then yeah throwing an execption is a good idea. But this means that the other model_exporter will blow up also. Btw, what will actually catch this exception and what would the handling be?